### PR TITLE
Feature/lab 3 letters

### LINKED ISSUE #456: Lab 03 - Letters
# Lab 3 - Letter frequency

- Create an executable go program in directory `03_letters/USERNAME`
- Write a function that returns a mapping of each letter to its frequency:

```
func letters(s string) map[rune]int
```

Write a function that returns a sorted slice of strings with elements `"{key}:{val}"`. Use package [sort](https://golang.org/pkg/sort/):

```
func sortLetters(m map[rune]int) []string
```

Call `fmt.Println(strings.Join(sortLetters(letters("aba"))), "\n")` in `main` to print:

```
a:2
b:1
```

Bonus points: comprehensive tests

### PR DESCRIPTION
**Pull request by starfallz**
Fixes #456

Review of colleague's PR #

#### Changes proposed in this PR:

- Write a function that returns a mapping of each letter to its frequency

### CODE REVIEW

File: 03_letters/starfallz/main_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+func TestSortRuneFunction(t *testing.T) {
+	testCases := []struct {
+		description string
+		input       []rune
+		expected    []rune
+	}{
+		{"Test sort unsorted rune slice to sorted rune slice", []rune{97, 82}, []rune{82, 97}},
+		{"Test sort negative rune", []rune{-97, 82}, []rune{-97, 82}},
```
Comment (n0npax) [inline 03_letters/starfallz/main_test.go]: I would prefer to keep rune as rune: i.e. `e` instead of 99

File: 03_letters/starfallz/main.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	for _, currentLetter := range s {
+		letterCount[currentLetter]++
+	}
+
+	return letterCount
+}
+
+func sortLetters(m map[rune]int) []string {
+	allKeys := make([]rune, 0, len(m))
```
Comment (n0npax) [inline 03_letters/starfallz/main.go:29]: I believe `len: 0` and `cap: len(m)` should be replaced with  `len: len(m)`. You don't plan to grow slice, you already know the size.

File: 03_letters/starfallz/main_test.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+func TestSortLettersFunction(t *testing.T) {
+	testCases := []struct {
+		description string
+		input       map[rune]int
+		expected    []string
+	}{
+		{"Test sort letters using letters() as input and return count", letters("aba"), []string{"a:2", "b:1"}},
+		{"Test sort letters with uppercase using letters() as input and return count", letters("Aba"),
+			[]string{"A:1", "a:1", "b:1"}},
+		{"Test sort letters with uppercase and return count", map[rune]int{97: 2, 65: 3}, []string{"A:3", "a:2"}},
```
Comment (n0npax) [inline 03_letters/starfallz/main_test.go:62]: i would prefer to keep rune as rune (i.e. `'a':2 `instead of `97:2`

**Review by n0npax — CHANGES_REQUESTED** [review]
good work mate. Just 2 minor issues.

File: 03_letters/starfallz/main_test.go (hunk 3 above)
Comment (starfallz) [reply 03_letters/starfallz/main_test.go:62]: Cool no worries will change them. It's easier to read too. Thanks

File: 03_letters/starfallz/main.go (hunk 2 above)
Comment (starfallz) [reply 03_letters/starfallz/main.go:29]: Yeah you are right I am not sure why I did this. I will try it and see if that works or not. Thanks

File: 03_letters/starfallz/main.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func letters(s string) map[rune]int {
+	letterCount := make(map[rune]int)
+
+	for _, currentLetter := range s {
+		letterCount[currentLetter]++
+	}
+
+	return letterCount
```
Comment (alextmz) [inline 03_letters/starfallz/main.go:25]: Minor nitpick: on small functions like these, and between blocks that have just 1 line (the return),  internal line spacing doesn't add legibility, to the contrary. What do you think?

File: 03_letters/starfallz/main.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		allKeys = append(allKeys, k)
+	}
+
+	sortRune(allKeys)
+
+	result := make([]string, len(allKeys))
+
+	for i, currentKey := range allKeys {
+		result[i] = strings.Join([]string{string(currentKey), strconv.Itoa(m[currentKey])}, ":")
+	}
```
Comment (alextmz) [inline 03_letters/starfallz/main.go:40]: As a suggestion I'd try to do it without creating an intermediate slice - that's how I did it, but not yet reviewed, so not sure if it's the right way to go but I do think so. Makes things a lot shorter too.

File: 03_letters/starfallz/main_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		main()
+
+		expected := strconv.Quote(`a:2
+b:1`)
+		actual := strconv.Quote(buf.String())
+
+		if expected != actual {
+			t.Errorf("Unexpected output, expected: %s, actual: %s", expected, actual)
+		}
+	})
```
Comment (alextmz) [inline 03_letters/starfallz/main_test.go:24]: I'm not usually very big on getting things formatted properly but it definitely looks weird with the long line plus the extra spacing plus the line break for the literals. 

Also, don't capitalize error strings throughout (from https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments, see also https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: 03_letters/starfallz/main_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+func TestLettersFunction(t *testing.T) {
+	testCases := []struct {
+		description string
+		input       string
+		expected    map[rune]int
+	}{
+		{"Test letters to return sorted map of letters based on rune # with count", "abaabazzzzzz",
+			map[rune]int{97: 4, 98: 2, 122: 6}},
+		{"Test letters to handle uppercase letters", "aAAaA", map[rune]int{65: 3, 97: 2}},
+		{"Test letters to handle non-alphabetical characters", "    !!!*", map[rune]int{32: 4, 33: 3, 42: 1}},
```
Comment (alextmz) [inline 03_letters/starfallz/main_test.go:36]: Nice one testing punctuation and white spaces, I'd suggest also testing Unicode (eg. Emojis and whatanot). There's a whole bunch of weird characters out there.

File: 03_letters/starfallz/main_test.go (hunk 3 above)
Comment (alextmz) [reply 03_letters/starfallz/main_test.go:62]: Would make it much easier to read the weird Unicode too. By definition Go source code is UTF-8 encoded, so you shouldn't be afraid of putting weird characters in, every tool used should either support it or be abandoned.

File: 03_letters/starfallz/main_test.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+		expected := testCase.expected
+		t.Run(testCase.description, func(t *testing.T) {
+			sortRune(actual)
+
+			if !reflect.DeepEqual(actual, expected) {
+				t.Errorf("Unexpected output, expected: %d, actual: %d", expected, actual)
+			}
+		})
+	}
+}
```
Comment (alextmz) [inline 03_letters/starfallz/main_test.go:103]: The TestSortLettersFunction already tests sortRune indirectly, so I'd suggest moving the tests from TestSortRuneFunction to the former to simplify things.

**Review by alextmz — CHANGES_REQUESTED** [review]
The Uppercase error strings should be fixed, it is a nice habit to have that pays over time. I'd also suggest using characters directly instead of their code, as it makes things more legible (IMHO)

**Comment by juliaogris** [comment]
The go-course is now closed. Thank you very much for participating.
